### PR TITLE
[Merged by Bors] - Altair consensus changes and refactors

### PR DESCRIPTION
**Pull request by michaelsproul**
## Proposed Changes

Implement the consensus changes necessary for the upcoming Altair hard fork.

## Additional Info

This is quite a heavy refactor, with pivotal types like the `BeaconState` and `BeaconBlock` changing from structs to enums. This ripples through the whole codebase with field accesses changing to methods, e.g. `state.slot` => `state.slot()`.

### CODE REVIEW

**Comment by michaelsproul** [comment]
Current status:

- `make` compiles a Lighthouse binary but it doesn't yet work (it should be close)
- Actual Altair consensus code is stubbed out
- Many tests are broken due to the refactor
- Remaining tasks are tagged `FIXME(altair)` in the code

**Comment by michaelsproul** [comment]
https://github.com/sigp/lighthouse/issues/2278 should likely use this branch as a base. We could keep this branch clean with PRs + rebases, and then merge it without squashing. Or we could split consensus/networking changes into two PRs, and merge them one after the other.

**Comment by michaelsproul** [comment]
Lighthouse can now run against mainnet/pyrmont from this branch, I've temporarily hardcoded the fork slot to INT_MAX.

Will work on the Altair EF tests + consensus tomorrow.

**Comment by michaelsproul** [comment]
TODO:

- Change `SignedBeaconBlock` to a `superstruct` so that it's possible to hint which variant we want manually. This is necessary for @pawanjay176's networking changes.

Also a note for later:

- We'll need something analogous to the op pool to collect `SignedContributionAndProof`s from the network and aggregate them for use in proposed blocks. This should likely live on the `BeaconChain`. Spec here: https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.2/specs/altair/validator.md#preparing-a-beaconblock

File: consensus/types/src/signed_beacon_block.rs
Comment (michaelsproul) [inline consensus/types/src/signed_beacon_block.rs]: Note to self: this doesn't need the `match` any more

**Comment by michaelsproul** [comment]
It's looking good! All the consensus tests from the EF v1.1.0-alpha.3 are passing now.

The upcoming changes I have in mind are:

- [x] Review + merge Pawan's networking changes
- [x] Pools for `SignedContributionAndProof`s and other sync aggregate objects that can come off the network. See [Altair validator spec](https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.3/specs/altair/validator.md). This will likely be similar to the naive aggregation pool (and maybe the op pool), accepting items from the network via methods on the `BeaconChain`.
- [x] Abstract block proposal to propose Altair blocks (using above pool)
- [x] Adapt attestation packing in op pool to use Altair rewards for calculations (where relevant)
- [x] Annihilate those `TestingXBuilder`s. I reckon we throw them in the bin (and those fragile tests based on them). IMO the way forward is to use the `BeaconChainHarness` to generate attestations/slashings/etc for testing, because it has a lot more context to make that ergonomic + sensible. In other places where the existing builders are used, I think we can probably find alternatives that are less cumbersome (use `TestRandom` or use the `BeaconChainHarness`).
- [x] Get the `beacon_chain` tests passing (see previous)
- [x] Update `superstruct` to generate _partial getters_ for fields shared by some variants. I think this is preferable to the `as_base` and `as_altair` methods we use at the moment, because it will continue to work for future forks retaining the same fields (i.e. the merge). I've started working on this locally and will push to `sigp/superstruct` soon (TM).
- [x] Resolve the `ValidatorStatus` conundrum (see comments in `per_epoch_processing::altair`) [assigned: @michaelsproul]
- [x] Add a service to the validator client to perform sync committee duties. Requires the API from the next item. See [Altair validator spec](https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.3/specs/altair/validator.md#sync-committees).
- [x] Add APIs to the beacon node for fetching & broadcasting sync committee signatures and aggregates. We should check if a standard API has been proposed, and consider proposing one if not. Forming aggregates will depend on the pool(s) for `SignedContributionAndProof`s (item #2).
- [x] Plug in `upgrade_to_altair`
- [x] Add EF test runner for `upgrade_to_altair` (`fork` subdir)
- [x] Parametrise updated tests (`beacon_chain`, etc) by fork, possibly using env var approach for simplicity
- [x] Update `lcli` commands (`new_testnet` could be useful for running local Altair testnets soon)
- [ ] Update aggregate attestation gossip filter: https://github.com/ethereum/eth2.0-specs/pull/2225
- [x] Make sure to update or replace `BeaconChain::duration_to_next_fork` (spotted by Pawan)

Stretch goals:

- [ ] Implement runners for the EF fork choice tests (haven't looked into this).
- [x] Do some profiling/benchmarking and consider stripping out/caching proposer indices & total balances (links in with validator status considerations)
- [ ] Try supercharging tree hashing by wrapping `validators_mut()` so that it records the dirty indices.
- [ ] Strip all `/// Spec vX.Y.Z` comments (they're just a maintenance burden, and hugely out of date)

I've probably forgotten a few things, if anyone notices any, please post here!

**Comment by michaelsproul** [comment]
I've updated the "roadmap"/TODO list above.

I have an idea for how to resolve the validator status thing, so I'll tentatively take that for tomorrow. The idea is: translate the _old_ validator status into _new_ participation bits, and re-jig the validator monitor to mostly use the new stuff (with `legacy: Option<ValidatorStatus>` in there as well until the fork happens).

**Comment by michaelsproul** [comment]
Draft API spec from JGM that we can provide feedback on in the #apis channel of Eth R&D: https://hackmd.io/@QYHAVYiHRg65pdI_CEm7Eg/syncommitteeapi

**Comment by michaelsproul** [comment]
Need to mark `state_transition` tests as `release` only

We've also added ~10-15 mins to the release tests. I think it may make sense to split it into `beacon_chain` and non-`beacon_chain` tests (also so we can run the beacon_chain tests for multiple fork configs)

**Comment by michaelsproul** [comment]
I noticed we introduced some new indexing into vecs during Altair. I might add a clippy lint against this and take out all the indexing, depending on how much work that seems to be: https://rust-lang.github.io/rust-clippy/stable/index.html#indexing_slicing

**Comment by michaelsproul** [comment]
This is ready for review, modulo me getting CI to pass in a more timely fashion. Unfortunately my idea of splitting the tests was a bit of a non-starter, because most of the time (49 minutes) is spent on compilation!

Will keep investigating

**Comment by michaelsproul** [comment]
Commit c89de3ba6d9f232aa51c51f62ce81aef044583f4 is running on the SigP Pyrmont validators. I've also tested that commit with Infura's Pyrmont node, and it is working well (the Teku BN doesn't report any Altair-related fields on Pyrmont :+1: )

**Comment by michaelsproul** [comment]
The sync committee cache needs to removed in light of https://github.com/ethereum/eth2.0-specs/issues/2392, but that should be a fairly self-contained change and I think this PR is still otherwise ready for review.

**Comment by paulhauner** [comment]
I'm starting review here, *finally* :tada: 

It's a thicc boi so I'm going to do a few passes. On these initial passes I'll just leave some comments (instead of a "review"). Hopefully it's not too noisy!

File: consensus/types/src/beacon_state.rs
Comment (paulhauner) [inline consensus/types/src/beacon_state.rs]: I noticed that this exists in `Base`, not just `Altair`. Is that intentional?

Comment (michaelsproul) [reply consensus/types/src/beacon_state.rs]: It was for simplicity, but this entire cache will be deleted because it turns out to be impossible to implement correctly: ethereum/eth2.0-specs#2392

I'll do that today and push an update

File: consensus/types/src/beacon_state.rs
Comment (paulhauner) [inline consensus/types/src/beacon_state.rs:304]: I noticed that `BeaconState` doesn't seem to have a `new` constructor for Altair whilst `BeaconBlock` does:

https://github.com/sigp/lighthouse/blob/1a9c89944dd06b7976a5e10bc4c8697877cea713/consensus/types/src/beacon_block.rs#L60-L66

I'm not far enough in to understand the ramifications, but I'm just making a note of it for now.

Comment (michaelsproul) [reply consensus/types/src/beacon_state.rs:304]: The reason it's currently like this is that we were creating genesis states for Altair by first making a phase0 state and then upgrading it immediately using `upgrade_to_altair`. This was a bit unwieldy, and the spec has since been updated to clarify genesis initialisation for Altair-only nets: https://github.com/ethereum/eth2.0-specs/pull/2323

When we implement that update from the spec we may well need a `BeaconState::new` that works for Altair

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs
Comment (paulhauner) [inline beacon_node/eth2_libp2p/src/rpc/protocol.rs]:
```suggestion
        BeaconBlock::empty(&MainnetEthSpec::default_spec()),
```

I checked this locally and I'm confident it'll compile.

File: testing/ef_tests/src/handler.rs
Comment (paulhauner) [inline testing/ef_tests/src/handler.rs:262]:
```suggestion
        // requiring real crypto, so only run these tests with fake crypto for now.
```

Comment (michaelsproul) [reply testing/ef_tests/src/handler.rs:262]: The comment is right, these tests can only be run with real crypto because they do real signature aggregation to compute sync committees. The `cfg` is `not(feature = "fake-crypto")`, i.e. real.

https://github.com/sigp/lighthouse/blob/967d680a9801bcb83676f8cdb6825b23580148cc/testing/ef_tests/src/handler.rs#L261-L263

Comment (michaelsproul) [reply testing/ef_tests/src/handler.rs:262]: Ohh I just saw the ambiguous interpretation along the lines of

"we don't _have_ to use real crypto, so use fake"

when what was meant was:

"we need to use real crypto but the tests don't state that, so temporarily hardcode it"

Comment (michaelsproul) [reply testing/ef_tests/src/handler.rs:262]: I reckon we leave it as-is for now and see how things change in alpha.4 (I'll open an issue on the spec if it's not fixed)

File: consensus/types/src/sync_committee.rs
Comment (paulhauner) [inline consensus/types/src/sync_committee.rs]:
```suggestion
    /// Create a temporary sync committee that should *never* be included in a legitimate consensus object.
```

File: consensus/types/src/chain_spec.rs
Comment (paulhauner) [inline consensus/types/src/chain_spec.rs]:
```suggestion
//! Arguably the most important of these is `ChainSpec`, which is used throughout Lighthouse as the
```

File: consensus/types/src/consts.rs
Comment (paulhauner) [inline consensus/types/src/consts.rs]: I notice that `INACTIVITY_SCORE_BIAS` and `INACTIVITY_PENALTY_QUOTIENT_ALTAIR` are present here and in `ChainSpec` but I'm yet to figure out why.

Comment (michaelsproul) [reply consensus/types/src/consts.rs]: ooh I don't think there's any good reason for that, we can probably consolidate

File: beacon_node/beacon_chain/src/test_utils.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/test_utils.rs]: We use `pre_state` to refer to this state in other parts of the code (it also seems like a common term outside of LH too) so perhaps `make_block_return_pre_state` would be more consistent? Just a nit really.

File: consensus/ssz_derive/src/lib.rs
Comment (paulhauner) [inline consensus/ssz_derive/src/lib.rs:160]: Out of interest, is there a reason (apart from simplicity) why they must all be variable-sized rather than consistently variable/fixed-size?

Comment (paulhauner) [reply consensus/ssz_derive/src/lib.rs:160]: Hmm perhaps they don't even need to be consistent :thinking:

Comment (paulhauner) [reply consensus/ssz_derive/src/lib.rs:160]: > Hmm perhaps they don't even need to be consistent thinking

Oh no, `Encode::is_ssz_fixed_len()` won't allow for this.

File: beacon_node/operation_pool/src/lib.rs
Comment (paulhauner) [inline beacon_node/operation_pool/src/lib.rs]: This `_spec` is unused. Not a big deal, just pointing it out. Perhaps it's a future boi.

File: consensus/state_processing/src/common/get_attestation_participation.rs
Comment (paulhauner) [inline consensus/state_processing/src/common/get_attestation_participation.rs]:
```suggestion
        && state.slot() == data.slot.safe_add(spec.min_attestation_inclusion_delay)?
```

Fixes a slight difference from the [spec](https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.3/specs/altair/beacon-chain.md#modified-process_attestation).

I don't think my suggested change would *actually* make a change at runtime, since attestations from prior to the inclusion distance will be filtered out before hitting this function. That goes for both `per_block_processing` and `operation_pool.`

File: consensus/state_processing/src/common/get_attestation_participation.rs
Comment (paulhauner) [inline consensus/state_processing/src/common/get_attestation_participation.rs]: Why don't we check the source here? I haven't been able to figure it out yet, perhaps it's something do with the op pool?

If it's just for efficiency, perhaps the gains aren't worth the risk of something "falling through the cracks"?

Comment (michaelsproul) [reply consensus/state_processing/src/common/get_attestation_participation.rs]: Yeah, you're right that it doesn't hurt to have that check in there. This function existed as a sort of weird intermediate that I had to yank out of attestation processing, but it's now been refactored in the spec into a real entity: https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.6/specs/altair/beacon-chain.md#get_attestation_participation_flag_indices

I'll fix it on this branch, and then carry that over when we do the update

File: consensus/ssz_derive/src/lib.rs
Comment (michaelsproul) [reply consensus/ssz_derive/src/lib.rs:160]: > Out of interest, is there a reason (apart from simplicity) why they must all be variable-sized rather than consistently variable/fixed-size?

Just simplicity. I figured all of our enums would be for variable-sized `superstructs`. I don't think it would be worth adding support for enums with equal fixed-size variants before we have some use for them (and I doubt we would).

File: consensus/state_processing/src/common/get_attestation_participation.rs
Comment (michaelsproul) [reply consensus/state_processing/src/common/get_attestation_participation.rs]: Resolved in https://github.com/sigp/lighthouse/pull/2279/commits/6415962290a54c00289cb526aefc5a304e9e94e4

File: consensus/state_processing/src/common/get_attestation_participation.rs
Comment (michaelsproul) [reply consensus/state_processing/src/common/get_attestation_participation.rs]: Resolved in https://github.com/sigp/lighthouse/pull/2279/commits/6415962290a54c00289cb526aefc5a304e9e94e4

File: beacon_node/operation_pool/src/lib.rs
Comment (michaelsproul) [reply beacon_node/operation_pool/src/lib.rs]: Fixed in https://github.com/sigp/lighthouse/pull/2279/commits/6415962290a54c00289cb526aefc5a304e9e94e4

File: beacon_node/beacon_chain/src/test_utils.rs
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/test_utils.rs]: Fixed in https://github.com/sigp/lighthouse/pull/2279/commits/6415962290a54c00289cb526aefc5a304e9e94e4

File: consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs
Comment (paulhauner) [inline consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs]:
```suggestion
    if (1..3).all(|i| bits.get(i).unwrap_or(false))
```

This is a non-substantive change, but it makes our code look more like the spec.

File: consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs
Comment (paulhauner) [inline consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs]:
```suggestion
    if (0..2).all(|i| bits.get(i).unwrap_or(false))
```

As per previous comment regarding looking more like the spec.

File: consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs
Comment (paulhauner) [reply consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs]: Addressed in https://github.com/sigp/lighthouse/pull/2367

File: consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs
Comment (paulhauner) [reply consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs]: Addressed in https://github.com/sigp/lighthouse/pull/2367

File: consensus/state_processing/src/per_epoch_processing/altair/participation_flag_updates.rs
Comment (paulhauner) [inline consensus/state_processing/src/per_epoch_processing/altair/participation_flag_updates.rs:16]: Hmm, looks like a bug in `rustfmt`. No change required, just pointing it out. I tried to put the `);` back onto L16 but `rustfmt` put it back :shrug:

File: consensus/types/src/participation_flags.rs
Comment (michaelsproul) [inline consensus/types/src/participation_flags.rs]: Just noticed this should be a `>=`. Fixing it in the alpha.5/6 branch, but can back port it as necessary.

File: consensus/types/src/participation_flags.rs
Comment (michaelsproul) [inline consensus/types/src/participation_flags.rs]: Off-by-one, as per above

File: consensus/state_processing/src/per_epoch_processing/altair/participation_flag_updates.rs
Comment (michaelsproul) [reply consensus/state_processing/src/per_epoch_processing/altair/participation_flag_updates.rs:16]: How weird!

**Review by paulhauner — APPROVED** [review]
After my longest and most detailed review *ever*, I'm very excited to add this little green tick. I spent quite a lot of time reviewing this code and I don't believe I found *any* consensus bugs. I did find interesting nuances and improvements so I'm satisfied that my review process was thorough. If there *are* bugs remaining, my approval means that I share in the responsibility equally. (Of course, no *legal* responsibility is assumed by *anyone* and we remain fully protected by our Apache 2.0 license).

Great work @michaelsproul and @realbigsean, this is an awesome effort. Not only did this include the Altair logic, it significantly improved our existing implementation through efforts like removing array-slicing. I got RSI by just *looking* at all the changes from direct field access to getter/setters. Very, very well done! :champagne: :clinking_glasses: 

As discussed with @michaelsproul, I think we should hold off on merging this until we cut another minor release. I also have [a PR](https://github.com/sigp/lighthouse/pull/2369) that's targeting this branch.

I suspect that we'll merge a few more things into this PR over time. We can do re-reviews as this happens.

**Comment by michaelsproul** [comment]
Will fix merge conflicts here shortly before merging (in order to reduce churn in all the downstream PRs)

**Review by paulhauner — APPROVED** [review]
Merge it!

**Comment by michaelsproul** [comment]
Can't quite believe it... let's try this

bors r+

:tada:

**Pull request closed without merging**
